clang/lib/StaticAnalyzer/Checkers/SimpleBoundsChecker.cpp

```diff
@@ -0,0 +1,462 @@
+//== SimpleBoundsChecker.cpp ------------------------------*- C++ -*--==//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines SimpleBoundsChecker, which is a path-sensitive checker
+// which looks for an out-of-bound accesses.
+// 
+// The two main differences compared to other bounds checkers are:
+// 1. This checker is Checked-C Bounds aware. It reads bounds-safe information
+//    on the function declaration and check the index against the bounds
+// 2. It uses Z3 for handling complex non-concrete bounds constraint. Clang
+//    should be compiled with the Z3 solver enabled.
+//
+// Debug Preprocessor Flags:
+// - DEBUG_DUMP: dumps memory regions, bounds and index expressions,
+//               generated SMT formulas, and some return on failure causes
+// - DEBUG_VISITORS: dumps visited nodes when traversing expressions
+//                   and SVals in replaceSVal and getSymExpr
+//===----------------------------------------------------------------------===//
+
+#include "clang/StaticAnalyzer/Checkers/BuiltinCheckerRegistration.h"
+#include "clang/StaticAnalyzer/Core/BugReporter/BugType.h"
+#include "clang/StaticAnalyzer/Core/Checker.h"
+#include "clang/StaticAnalyzer/Core/CheckerManager.h"
+#include "clang/StaticAnalyzer/Core/PathSensitive/CheckerContext.h"
+#include "clang/StaticAnalyzer/Core/PathSensitive/ExprEngine.h"
+#include "clang/StaticAnalyzer/Core/PathSensitive/SMTConv.h"
+#include "clang/StaticAnalyzer/Core/PathSensitive/SMTSolver.h"
+
+#include "clang/AST/ASTConsumer.h"
+#include "clang/AST/RecursiveASTVisitor.h"
+#include "clang/StaticAnalyzer/Core/PathSensitive/SValVisitor.h"
+
+#include <string>
+
+#define DEBUG_DUMP 0
+#define DEBUG_VISITORS 0
+
+using namespace clang;
+using namespace ento;
+
+namespace {
+  class SimpleBoundsChecker : public Checker<check::Location> {
+    mutable std::unique_ptr<BuiltinBug> BT;
+
+    void checkBoundsInfo(const DeclaratorDecl* decl, std::string label, ASTContext& Ctx) const;
+
+    // Replaces the symbol 'from' with the symbol 'to' in the symbolic expression 'E'
+    SVal replaceSVal(ProgramStateRef state, SVal E, SVal from, SVal to) const;
```

Comment:
Please make the variable names more uniform throughout this PR. I see a mix of camelCase and PascalCase as well as all lowercase variable names.